src/iquery.cpp

```diff
@@ -0,0 +1,167 @@
+
+    // clang-format off
+    desc.add_options()
+        ("columns,c", po::bool_switch(), "")
+        ("query_string", po::value<std::string>()->default_value("-"), "")
+        ("sql-only", po::bool_switch(), "")
+        ("zone,z", po::value<std::string>(), "")
+        ("help,h", "");
+    // clang-format on
+
+    po::positional_options_description pod;
+    pod.add("query_string", 1);
+
+    load_client_api_plugins();
+
+    try {
+        po::variables_map vm;
+        po::store(po::command_line_parser(_argc, _argv).options(desc).positional(pod).run(), vm);
+        po::notify(vm);
+
+        if (vm.count("help")) {
+            print_usage_info();
+            return 0;
+        }
+
+        if (vm["columns"].as<bool>()) {
+            print_columns_info();
+            return 0;
+        }
+
+        GenQuery2Input input{};
+
+        if (vm.count("query_string") == 0) {
+            fmt::print(stderr, "error: Missing QUERY_STRING\n");
+            return 1;
+        }
+
+        auto query_string = vm["query_string"].as<std::string>();
+
+        // Read from stdin.
+        if ("-" == query_string) {
+            query_string.clear();
+            std::getline(std::cin, query_string);
+        }
+
+        if (query_string.empty()) {
+            fmt::print(stderr, "error: Missing QUERY_STRING\n");
+            return 1;
+        }
+
+        input.query_string = query_string.data();
+
+        std::string zone;
+        if (vm.count("zone")) {
+            zone = vm["zone"].as<std::string>();
+            input.zone = zone.data();
+        }
+
+        if (vm["sql-only"].as<bool>()) {
+            input.sql_only = 1;
+        }
+
+        irods::experimental::client_connection conn;
+        char* sql{};
+        irods::at_scope_exit free_sql{[&sql] {
+            if (sql) {
+                std::free(sql);
+            }
+        }};
+
+        const auto ec = rc_genquery2(static_cast<RcComm*>(conn), &input, &sql);
+        if (ec != 0) {
+            fmt::print(stderr, "error: {}\n", ec);
+            return 1;
+        }
+
+        fmt::print("{}\n", sql);
+
+        return 0;
+    }
+    catch (const irods::exception& e) {
+        fmt::print(stderr, "error: {}\n", e.client_display_what());
+    }
+    catch (const std::exception& e) {
+        fmt::print(stderr, "error: {}\n", e.what());
+    }
+
+    return 1;
+} // main
+
+auto print_usage_info() -> void
+{
+	fmt::print(R"_(iquery - Query the iRODS Catalog
+
+Usage: iquery [OPTION]... QUERY_STRING
+
+Queries the iRODS Catalog using GenQuery2.
+
+QUERY_STRING is expected to be a string matching the GenQuery2 syntax. Failing
+to meet this requirement will result in an error.
+
+If QUERY_STRING is a hyphen (-) or is empty, input is read from stdin. Input
+taken via stdin will be viewed as the QUERY_STRING to execute. For example:
+
+    echo select COLL_NAME, DATA_NAME | iquery
+
+Mandatory arguments to long options are mandatory for short options too.
```

Comment:
looking at the options... what does this mean for this icommand?

can this line be removed?

Reply:
That line is primarily about the `-z` option.

It was modeled after other POSIX commands. The man page for `ls` uses those exact words.

It's not required, but it was included so that it's clear to users that `-z` takes the same argument as `--zone`.

Reply:
right, `zone` is the only option that takes a parameter.

the use of 'mandatory' and that it has a default value is what threw me off a little.  but it is not confusing, just seemed ... superfluous.

definitely fine to leave it as is.

Reply:
Cool. Leaving as is.